firmware/library/L4_Application/task_scheduler.cpp

```diff
@@ -0,0 +1,109 @@
+        TaskScheduler::Instance().GetPreRunEventGroupHandle();
+    const EventBits_t kPreRunSyncBits =
+        TaskScheduler::Instance().GetPreRunSyncBits();
+    const uint32_t kSyncBit = (1 << kTaskIndex);
+    SJ2_ASSERT_FATAL(task.PreRun(),
+                     "PreRun() failed for task: %s, terminating scheduler!",
+                     task.GetName());
+    // wait for all other PreRun() of other tasks to finish
+    xEventGroupSync(pre_run_event_group_handle, kSyncBit, kPreRunSyncBits,
+                    portMAX_DELAY);
+    // ALl PreRun() complete, each Task's Run() can now start executing...
+    TickType_t last_wake_time = xTaskGetTickCount();
+    while (1)
+    {
+        if (!task.Run())
+        {
+            SJ2_ASSERT_WARNING(
+                false,
+                "An error occured, the following task will be suspended: %s",
+                task.GetName());
+            vTaskSuspend(NULL);
+        }
+        // delay task if the task's delay time has been set...
+        uint32_t delay_time = task.GetDelayTime();
+        if (delay_time)
+        {
+            vTaskDelayUntil(&last_wake_time, delay_time);
+        }
+    }
+};
+void rtos::TaskScheduler::Start()
+{
+    SJ2_ASSERT_FATAL(
+        task_count_ != 0,
+        "Attempting to start TaskScheduler but there are no tasks that "
+        "are currently scheduled.");
+    InitializeAllTasks();
+    vTaskStartScheduler();
+};
+
+void rtos::TaskScheduler::InitializeAllTasks()
+{
+    for (uint32_t i = 0; i < task_count_; i++)
+    {
+        TaskInterface * task = task_list_[i];
+        *(task->GetHandle()) = xTaskCreateStatic(
+            RunTask, task->GetName(),
+            static_cast<uint16_t>(StackSize(task->GetStackSize())),
+            PassParameter(task), task->GetPriority(), task->GetStack(),
+            task->GetTaskBuffer());
+        SJ2_ASSERT_FATAL(task->GetHandle() != nullptr,
+                         "Unable to create task: %s", task->GetName());
+        SJ2_ASSERT_FATAL(task->Setup(),
+                         "Failed to complete Setup() for task: %s",
+                         task->GetName());
+        pre_run_sync_bits_ |= (1 << i);
+    }
+    pre_run_task_handle_ = task_list_[0]->GetHandle();
+    pre_run_event_group_handle_ =
+        xEventGroupCreateStatic(&pre_run_event_group_buffer_);
+    SJ2_ASSERT_FATAL(pre_run_event_group_handle_ != nullptr,
+                     "Failed to create PreRun Event Group!");
+};
+
+void rtos::TaskScheduler::RemoveTask(const char * task_name)
+{
+    const uint8_t kTaskIndex = GetTaskIndex(task_name);
+    if (kTaskIndex > task_count_)
+    {
+        return;
+    }
+    vTaskDelete(task_list_[kTaskIndex]->GetHandle());
+    // remove the task and update the task_list_ and task_count_
+    for (uint8_t i = kTaskIndex; i < task_count_; i++)
+    {
+        task_list_[i] = task_list_[i + 1];
+    }
+    task_count_--;
+    task_list_[task_count_] = nullptr;
+};
```

Comment:
So, this implementation is fine. But I don't think we need to re-order the list each time something is removed. We could simply fill up task array items that are set to `nullptr`. If all tasks spaces are defined, then we know the task list is full.